may-2018-reenabled-opcodes.md

```diff
@@ -0,0 +1,428 @@
+
+Impact of successful execution:
+* stack memory use reduced by `len(x1)`
+* number of elements on stack is reduced by one
+
+Unit tests:
+1. `x1 x2 OP_XOR -> failure`, where `len(x1) != len(x2)`. The two operands must be the same size.
+2. `x1 x2 OP_XOR -> x1 xor x2`. Check valid results.
+    
+## Arithmetic
+
+#### Note about canonical form and floor division
+
+Operands for all arithmetic operations are assumed to be numeric values and must be in canonical form.  
+See [data types](#data-types) for more information.
+
+**Floor division**
+
+Note: that when considering integer division and modulo operations with negative operands the rules applied in the C language and most
+languages (with Python being a notable exception) differ from the strict mathematical definition.  Script follows the C language set of
+rules.  Namely:
+1. Non-integer quotients are rounded towards zero
+2. The equation `(a/b)*b + a%b == a` is satisfied by the results
+3. From the above equation it follows that: `a%b == a - (a/b)*b`
+4. In practice if `a` is negative for the modulo operator the result will be negative or zero.
+
+
+### OP_DIV
+
+    Opcode (decimal): 150
+    Opcode (hex): 0x96
+    
+Return the integer quotient of `a` and `b`.  If the result would be a non-integer it is rounded *towards* zero.
+
+    a b OP_DIV -> out
+    
+    where a and b are interpreted as numeric values
+    
+The operator must fail if:
+1. `!isnum(a) || !isnum(b)`. Fail if either operand is not a numeric value.
+1. `b == 0`. Fail if `b` is equal to any type of zero.
+
+Impact of successful execution:
+* stack memory use reduced
+* number of elements on stack is reduced by one
+
+Unit tests:
+1. `a b OP_DIV -> failure` where `!isnum(a)` or `!isnum(b)`. Both operands must be numeric values
+2. `a 0 OP_DIV -> failure`. Division by positive zero (all sizes), negative zero (all sizes), `OP_0` 
+4. `27 7 OP_DIV -> 3`, `27 -7 OP_DIV -> -3`, `-27 7 OP_DIV -> -3`, `-27 -7 OP_DIV -> 3`. Check negative operands. 
+  *Pay attention to sign*.
+5. check valid results for operands of different lengths `0..4`
+    
+### OP_MOD
+
+    Opcode (decimal): 151
+    Opcode (hex): 0x97
+
+Returns the remainder after dividing a by b.  The output will be represented using the least number of bytes required. 
+
+	a b OP_MOD -> out
+	
+	where a and b are interpreted as numeric values
+	
+The operator must fail if:
+1. `!isnum(a) || !isnum(b)`. Fail if either operand is not a numeric value.
+1. `b == 0`. Fail if `b` is equal to any type of zero.
+
+Impact of successful execution:
+* stack memory use reduced (one element removed)
+* number of elements on stack is reduced by one
+
+Unit tests:
+1. `a b OP_MOD -> failure` where `!isnum(a)` or `!isnum(b)`. Both operands must be numeric values.
+2. `a 0 OP_MOD -> failure`. Division by positive zero (all sizes), negative zero (all sizes), `OP_0` 
+4. `27 7 OP_MOD -> 6`, `27 -7 OP_MOD -> 6`, `-27 7 OP_MOD -> -6`, `-27 -7 OP_MOD -> -6`. Check negative operands. 
```

Comment:
Test case 3 is missing. Either renumber the subsequent ones, or add a test case for zero-result cases (e.g., `27 3 OP_MOD -> 0`, etc.)

Reply:
markdown renumbers automatically, but updated. Zero result cases added, thanks.